tools/clang/lib/AST/HlslTypes.cpp

```diff
@@ -704,32 +664,27 @@ bool IsHLSLObjectWithImplicitROMemberAccess(clang::QualType type) {
 }
 
 bool IsHLSLRWNodeInputRecordType(clang::QualType type) {
-  if (const RecordType *RT = type->getAs<RecordType>()) {
-    StringRef name = RT->getDecl()->getName();
-    if (name == "RWDispatchNodeInputRecord" ||
-        name == "RWGroupNodeInputRecords" || name == "RWThreadNodeInputRecord")
-      return true;
-  }
+  if (const HLSLNodeObjectAttr *Attr = getNodeAttr(type))
+    return Attr->getType() == HLSLNodeObjectAttr::RWDispatchNodeInputRecord ||
+           Attr->getType() == HLSLNodeObjectAttr::RWGroupNodeInputRecords ||
+           Attr->getType() == HLSLNodeObjectAttr::RWThreadNodeInputRecord;
   return false;
 }
 
 bool IsHLSLRONodeInputRecordType(clang::QualType type) {
-  if (const RecordType *RT = type->getAs<RecordType>()) {
-    StringRef name = RT->getDecl()->getName();
-    if (name == "DispatchNodeInputRecord" || name == "GroupNodeInputRecords" ||
-        name == "ThreadNodeInputRecord")
-      return true;
-  }
+  if (const HLSLNodeObjectAttr *Attr = getNodeAttr(type))
+    return Attr->getType() == HLSLNodeObjectAttr::DispatchNodeInputRecord ||
+           Attr->getType() == HLSLNodeObjectAttr::GroupNodeInputRecords ||
+           Attr->getType() == HLSLNodeObjectAttr::ThreadNodeInputRecord;
   return false;
 }
 
 bool IsHLSLNodeOutputType(clang::QualType type) {
-  if (const RecordType *RT = type->getAs<RecordType>()) {
-    StringRef name = RT->getDecl()->getName();
-    if (name == "NodeOutput" || name == "NodeOutputArray" ||
-        name == "EmptyNodeOutput" || name == "EmptyNodeOutputArray")
-      return true;
-  }
+  if (const HLSLNodeObjectAttr *Attr = getNodeAttr(type))
+    return Attr->getType() == HLSLNodeObjectAttr::NodeOutput ||
+           Attr->getType() == HLSLNodeObjectAttr::NodeOutputArray ||
+           Attr->getType() == HLSLNodeObjectAttr::EmptyNodeOutput ||
+           Attr->getType() == HLSLNodeObjectAttr::EmptyNodeOutputArray;
```

Comment:
```c++
  return (static_cast<uint32_t>(GetNodeIOType(type)) &
          static_cast<uint32_t>(DXIL::NodeIOFlags::Output)) != 0;
```

Reply:
I believe this needs be:
```cpp
  return (static_cast<uint32_t>(GetNodeIOType(type)) &
          (static_cast<uint32_t>(DXIL::NodeIOFlags::Output) |
           static_cast<uint32_t>(DXIL::NodeIOFlags::RecordGranularityMask))) ==
         static_cast<uint32_t>(DXIL::NodeIOFlags::Output);
```
To avoid the output record type, since this should only return true if it's an output node object type.

Reply:
I think this loses a lot of readability...